src/object/cli_shard.c

```diff
@@ -749,10 +749,17 @@ dc_rw_cb(tse_task_t *task, void *arg)
 			D_DEBUG(DB_IO, DF_UOID" size "DF_U64
 				" eph "DF_U64"\n", DP_UOID(orw->orw_oid),
 				sizes[i], orw->orw_epoch);
-			if (!is_ec_obj || reasb_req->orr_fail == NULL ||
-			    iods[i].iod_size == 0 || sizes[i] != 0)
+			if (!is_ec_obj) {
 				iods[i].iod_size = sizes[i];
-			if ((is_ec_obj && reasb_req->orr_recov) &&
+				continue;
+			}
+
+			if (!reasb_req->orr_size_set || iods[i].iod_size == 0) {
```

Comment:
seems need to add "|| sizes[i] != 0" to L757 or L752.
For the case that user set incorrect iod_size in fetch req, and 2 shards returns size as 0 and non-zero (correct value).

Reply:
hmm, i did a little test, in my test  after this patch the "daos_test -x -w ./io_conf/daos_io_conf_4" cannot pass.
And with below patch on your PR, it can work:
diff --git a/src/object/cli_shard.c b/src/object/cli_shard.c
index 7e424ef..5c5c362 100644
--- a/src/object/cli_shard.c
+++ b/src/object/cli_shard.c
@@ -754,7 +754,9 @@ dc_rw_cb(tse_task_t *task, void *arg)
                                continue;
                        }
 
-                       if (!reasb_req->orr_size_set || iods[i].iod_size == 0) {
+                       if (reasb_req->orr_fail == NULL &&
+                           (!reasb_req->orr_size_set ||
+                            iods[i].iod_size == 0)) {
                                iods[i].iod_size = sizes[i];
                                reasb_req->orr_size_set = 1;
                        }

We can see more tmr.

Reply:
ah, I missed that in the patch. sure, I will add it.